src/client/texturesource.cpp

```diff
@@ -1848,58 +1856,115 @@ bool TextureSource::generateImagePart(std::string_view part_of_name,
 
 #undef CHECK_DIM
 
-/*
-	Calculate the color of a single pixel drawn on top of another pixel.
 
-	This is a little more complicated than just video::SColor::getInterpolated
-	because getInterpolated does not handle alpha correctly.  For example, a
-	pixel with alpha=64 drawn atop a pixel with alpha=128 should yield a
-	pixel with alpha=160, while getInterpolated would yield alpha=96.
-*/
-static inline video::SColor blitPixel(const video::SColor src_c, const video::SColor dst_c, u32 ratio)
+namespace {
+
+/// Draw src on top of dst
+template <bool overlay, class Colour>
+void blit_pixel(const Colour &src, Colour &dst)
 {
-	if (dst_c.getAlpha() == 0)
-		return src_c;
-	video::SColor out_c = src_c.getInterpolated(dst_c, (float)ratio / 255.0f);
-	out_c.setAlpha(dst_c.getAlpha() + (255 - dst_c.getAlpha()) *
-		src_c.getAlpha() * ratio / (255 * 255));
-	return out_c;
+	if (src.a == 255 || dst.a == 0) {
+		if constexpr (overlay) {
+			if (dst.a != 255)
+				return;
+		}
+		// The top pixel is fully opaque or the bottom pixel is
+		// fully transparent -> replace the colour
+		dst = src;
+	} else if (src.a == 0) {
+		// A fully transparent pixel is on top -> do nothing
+		return;
+	} else if (dst.a == 255) {
+		// A semi-transparent pixel is on top and an opaque one in
+		// the bottom -> lerp r, g, and b
+		dst.r = (dst.r * (255 - src.a) + src.r * src.a) / 255;
+		dst.g = (dst.g * (255 - src.a) + src.g * src.a) / 255;
+		dst.b = (dst.b * (255 - src.a) + src.b * src.a) / 255;
+	} else {
+		if constexpr (overlay) {
+			return;
+		} else {
+			// A semi-transparent pixel is on top of a
+			// semi-transparent pixel -> general alpha compositing
+			auto a_new_255{src.a * 255 + (255 - src.a) * dst.a};
+			dst.r = (dst.r * (255 - src.a) * dst.a + src.r * src.a * 255)
+				/ a_new_255;
+			dst.g = (dst.g * (255 - src.a) * dst.a + src.g * src.a * 255)
+				/ a_new_255;
+			dst.b = (dst.b * (255 - src.a) * dst.a + src.b * src.a * 255)
+				/ a_new_255;
+			dst.a = a_new_255 / 255;
+		}
+	}
 }
 
-/*
-	Draw an image on top of another one, using the alpha channel of the
-	source image
-
-	This exists because IImage::copyToWithAlpha() doesn't seem to always
-	work.
-*/
-template<bool overlay>
-static void blit_with_alpha(video::IImage *src, video::IImage *dst,
-		v2s32 src_pos, v2s32 dst_pos, v2u32 size)
+/// A helper function for blit_with_alpha to support different endianesses
+template<bool overlay, class Colour>
+void blit_with_alpha_any_endian(video::IImage *src, video::IImage *dst,
+	const v2s32 &dst_pos, const v2u32 &size)
 {
-	auto src_dim = src->getDimension();
-	auto dst_dim = dst->getDimension();
+	if (dst->getColorFormat() != video::ECF_A8R8G8B8)
+		throw BaseException("blit_with_alpha() supports only ECF_A8R8G8B8 "
+			"destination images.");
+
+	auto src_dim{src->getDimension()};
+	auto dst_dim{dst->getDimension()};
+	bool drop_src{false};
```

Comment:
this is just my personal nitpick but what's the practical advantage of `{ }` over ` = ` here?

Reply:
With `{ }`, it is easier to distinguish between initialisations and assignments.
I think `{ }` is also more strict; for example, 
with ```int x{1.5};``` gcc shows an `error: narrowing conversion of '1.5e+0' from 'double' to 'int' [-Wnarrowing]` error message 
and with ```int x = 1.5;```, no error appears.

Reply:
Just mentioning that I also dislike this notation. Whereas the type checking is nice, it does make the code much less readable and in the case of `auto` types, it makes no sense at all.

Reply:
I have changed the code to avoid most brace initialisations.
Thanks for the advice on code style.